packages/metro-config/src/defaults/exclusionList.js

```diff
@@ -15,10 +15,19 @@ var list = [/website\/node_modules\/.*/, /.*\/__tests__\/.*/];
 
 function escapeRegExp(pattern) {
   if (Object.prototype.toString.call(pattern) === '[object RegExp]') {
-    return pattern.source.replace(/\//g, path.sep);
+    // the forward slash may or may not be escaped in regular expression depends
+    // on if it's in brackets. eg. /foo\/bar/ and /[/\\]foo/ are both valid.
+    // We should replace all forward slashes to proper OS specific separators.
+    // The separator needs to be escaped in the regular expression source string,
+    // hence the '\\' prefix.
+    return pattern.source.replace(/\/|\\\//g, '\\' + path.sep);
   } else if (typeof pattern === 'string') {
-    var escaped = pattern.replace(/[\-\[\]\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
-    // convert the '/' into an escaped local file separator
+    // escape back slashes in regular expression so that when
+    // it's used in RegExp constructor, the back slashes are preserved.
+    // var escaped = pattern.replace(/[\-\[\]\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&');
```

Comment:
Yes, I can add a test case for this. The special characters are fine when creating regular expressions, the only thing that needs to be escaped is the escape character (`\`) itself.

For example, the original logic escaped all special characters in the string, so that the input string
```
"[/\\]foo"
```
 which is effectively "[/\]foo" after string escape, will become
```
"\[\/\\\]foo"
```
It then replaced all forward slashes to the OS specific separators, let's say is `\\` for windows:
```
"\[\\\\\\]foo"
```
Now, we are going to use this to construct a regular expression instance. This is clearly a wrong string to use, as `new RegExp("\[\\\\\\]foo")` is the same to `/[\\\]foo/`, which is giving `Invalid regular expression` error.

My change here will result in the correct result with this steps:
`"[/\\]foo"` which is `"[/\]foo"` after string escape --> `"[/\\]foo"` --> `"[\\\\]foo"` --> `new RegExp("[\\\\]foo")` --> `/[\\]foo/`

Which is expected.